### PR TITLE
Port to C++20.

### PR DESCRIPTION
**Pull request by gatehouse**
This patch replaces various constructs that were deprecated in C++17 and removed in C++20 with their C++20 counterparts.

With this patch, s2geometry can be compiled with gcc 10.

### CODE REVIEW

**Comment by jmr** [comment]
What's the minimum c++ version needed for this to work?

**Comment by gatehouse** [comment]
> What's the minimum c++ version needed for this to work?
@jmr I would expect this to work from C++14 upwards.

**Comment by jmr** [comment]
This isn't compiling:

https://travis-ci.org/github/google/s2geometry/builds/696839883

C++20 is too new to require. C++14 is definitely ok and C++17 is probably ok.

**Comment by jmr** [comment]
> With this patch, s2geometry can be compiled with gcc 10.

gcc 10 works without this patch.
https://travis-ci.org/github/google/s2geometry/jobs/722137049#L423

File: src/s2/base/logging.h [hunk 1]
```diff
@@ -91,7 +91,7 @@ class S2FatalLogMessage : public S2LogMessage {
                     absl::LogSeverity severity, std::ostream& stream)
       ABSL_ATTRIBUTE_COLD
     : S2LogMessage(file, line, severity, stream) {}
-  ABSL_ATTRIBUTE_NORETURN ~S2FatalLogMessage() { abort(); }
+  ABSL_ATTRIBUTE_NORETURN ~S2FatalLogMessage() { throw std::runtime_error("s2geometry fatal error"); }
```
Comment (jmr) [inline src/s2/base/logging.h:94]: Please separate out changes unrelated to c++17-deprecated interfaces.

File: src/s2/s2shape_index.h [hunk 2]
```diff
@@ -272,8 +272,7 @@ class S2ShapeIndex {
   //   for (S2Shape* shape : index) { ... }
   //
   // CAVEAT: Returns nullptr for shapes that have been removed from the index.
-  class ShapeIterator
-      : public std::iterator<std::forward_iterator_tag, S2Shape*> {
```
Comment (jmr) [inline src/s2/s2shape_index.h:276]: This provided typedefs. Where do they come from now?

File: CMakeLists.txt [hunk 3]
```diff
@@ -13,7 +13,7 @@ if (APPLE)
     set(CMAKE_MACOSX_RPATH TRUE)
 endif()
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 20)
```
Comment (jmr) [inline CMakeLists.txt:16]: You can bump this to 17 if it's really needed to compile any of the replacements.

**Review by jmr — CHANGES_REQUESTED** [review]

**Comment by jmr** [comment]
I will close and re-open this in hope of getting travis-ci to notice the `.travis.yml` changes. The current `CMakeLists.txt` is incompatible with the old `.travis.yml`.

**Comment by jmr** [comment]
This breaks both compilers that support c++20
https://travis-ci.org/github/google/s2geometry/jobs/722649190#L544

and those that don't
https://travis-ci.org/github/google/s2geometry/jobs/722649187#L324

**Comment by jmr** [comment]
I think all C++20 problems should have been fixed.  If not, reopen a new issue/PR.

**Pull request closed without merging**